File: apps/mobile-app/src/services/webauthnService.ts
// src/services/webauthnService.ts

import authService from './authService';
import { PasskeyCreateRequest, PasskeyGetRequest } from 'react-native-passkey';

// Base64URL utility functions (critical for React Native Passkey compatibility)
const base64url = {
  toBase64: (base64url: string): string => {
    let base64 = base64url.replace(/-/g, '+').replace(/_/g, '/');
    // Add padding if needed
    const padding = base64.length % 4;
    if (padding) {
      base64 += '='.repeat(4 - padding);
    }
    return base64;
  },
  
  fromBase64: (base64: string): string => {
    return base64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '');
  }
};

// Request & Response interfaces
interface RegistrationStartRequest {
  userId: string;
  username: string;
  authenticatorName?: string;
}

interface RegistrationStartResponse {
  challenge: string;
  rpId: string;
  rpName: string;
  timeout: number;
  attestation: string;
  user: {
    id: string;
    name: string;
    displayName: string;
  };
  pubKeyCredParams: Array<{
    type: string;
    alg: number;
  }>;
  authenticatorSelection: {
    authenticatorAttachment: string;
    userVerification: string;
    requireResidentKey: boolean;
  };
}

interface RegistrationCompleteRequest {
  credentialId: string;
  credentialType: string;
  clientDataJSON: string;
  attestationObject: string;
  authenticatorName?: string;
}

interface AuthenticationStartRequest {
  username: string;
}

interface AuthenticationStartResponse {
  challenge: string;
  rpId: string;
  timeout: number;
  userVerification: string;
  allowCredentials: Array<{
    type: string;
    id: string;
    transports: string[];
  }>;
}

interface AuthenticationCompleteRequest {
  credentialId: string;
  credentialType: string;
  clientDataJSON: string;
  authenticatorData: string;
  signature: string;
  userHandle?: string;
}

interface AuthenticationResponse {
  token: string;
  userId: string;
  username: string;
  tokenType: string;
  expiresIn: number;
}

interface WebAuthnCredential {
  id: string;
  userId: string;
  userLabel: string;
  credentialData: string;
  secretData: string;
}

interface ApiResponse<T> {
  success: boolean;
  message: string;
  data?: T;
  error?: string;
}

class WebAuthnService {
  private baseUrl = '/webauthn';

  private convertToPasskeyCreateRequest(response: RegistrationStartResponse): PasskeyCreateRequest {
    return {
      // Convert challenge from base64url to base64 for react-native-passkey
      challenge: base64url.toBase64(response.challenge),
      rp: {
        id: response.rpId,
        name: response.rpName
      },
      user: {
        // Convert user ID from base64url to base64
        id: base64url.toBase64(response.user.id),
        name: response.user.name,
        displayName: response.user.displayName
      },
      pubKeyCredParams: response.pubKeyCredParams,
      authenticatorSelection: response.authenticatorSelection,
      timeout: response.timeout,
      attestation: response.attestation
    };
  }

  private convertToPasskeyGetRequest(response: AuthenticationStartResponse): PasskeyGetRequest {
    return {
      // Convert challenge from base64url to base64
      challenge: base64url.toBase64(response.challenge),
      rpId: response.rpId,
      allowCredentials: response.allowCredentials.map(cred => ({
        ...cred,
        // Convert credential ID from base64url to base64
        id: base64url.toBase64(cred.id)
      })),
      userVerification: response.userVerification,
      timeout: response.timeout
    } as PasskeyGetRequest; 
  }

  private convertRegistrationResponse(result: any): RegistrationCompleteRequest {
    return {
      // Convert all base64 values back to base64url for the backend
      credentialId: base64url.fromBase64(result.id),
      credentialType: result.type ?? "public-key",
      clientDataJSON: base64url.fromBase64(result.response.clientDataJSON),
      attestationObject: base64url.fromBase64(result.response.attestationObject),
    };
  }

  private convertAuthenticationResponse(result: any): AuthenticationCompleteRequest {
    return {
      // Convert all base64 values back to base64url for the backend
      credentialId: base64url.fromBase64(result.id),
      credentialType: result.type ?? "public-key",
      clientDataJSON: base64url.fromBase64(result.response.clientDataJSON),
      authenticatorData: base64url.fromBase64(result.response.authenticatorData),
      signature: base64url.fromBase64(result.response.signature),
      userHandle: result.response.userHandle ? base64url.fromBase64(result.response.userHandle) : undefined,
    };
  }

  async startRegistration(request: RegistrationStartRequest): Promise<PasskeyCreateRequest> {
    try {
      const response = await authService.apiRequest(`${this.baseUrl}/register/start`, {
        method: 'POST',
        body: JSON.stringify(request),
      });
      const data: ApiResponse<RegistrationStartResponse> = await response.json();
      
      if (data.success && data.data) {
        return this.convertToPasskeyCreateRequest(data.data);
      } else {
        throw new Error(data.error || 'Registration start failed');
      }
    } catch (error) {
      console.error('Error starting WebAuthn registration:', error);
      throw error;
    }
  }

  async completeRegistration(challenge: string, passkeyResult: any): Promise<{ message: string }> {
    try {
      // Convert the passkey result to the correct format
      const request = this.convertRegistrationResponse(passkeyResult);
      
      const response = await authService.apiRequest(`${this.baseUrl}/register/complete/${challenge}`, {
        method: 'POST',
        body: JSON.stringify(request),
      });
      const data: ApiResponse<string> = await response.json();
      
      if (data.success) {
        return { message: data.message };
      } else {
        throw new Error(data.error || 'Registration completion failed');
      }
    } catch (error) {
      console.error('Error completing WebAuthn registration:', error);
      throw error;
    }
  }

  async startAuthentication(request: AuthenticationStartRequest): Promise<PasskeyGetRequest> {
    try {
      const response = await authService.apiRequest(`${this.baseUrl}/authenticate/start`, {
        method: 'POST',
        body: JSON.stringify(request),
      });
      const data: ApiResponse<AuthenticationStartResponse> = await response.json();
      
      if (data.success && data.data) {
        return this.convertToPasskeyGetRequest(data.data);
      } else {
        throw new Error(data.error || 'Authentication start failed');
      }
    } catch (error) {
      console.error('Error starting WebAuthn authentication:', error);
      throw error;
    }
  }

  async completeAuthentication(challenge: string, passkeyResult: any): Promise<AuthenticationResponse> {
    try {
      // Convert the passkey result to the correct format
      const request = this.convertAuthenticationResponse(passkeyResult);
      
      const response = await authService.apiRequest(`${this.baseUrl}/authenticate/complete/${challenge}`, {
        method: 'POST',
        body: JSON.stringify(request),
      });
      const data: ApiResponse<AuthenticationResponse> = await response.json();
      
      if (data.success && data.data) {
        return data.data;
      } else {
        throw new Error(data.error || 'Authentication completion failed');
      }
    } catch (error) {
      console.error('Error completing WebAuthn authentication:', error);
      throw error;
    }
  }

  async getUserCredentials(userId: string): Promise<WebAuthnCredential[]> {
    try {
      const response = await authService.apiRequest(`${this.baseUrl}/credentials/${userId}`);
      const data: ApiResponse<WebAuthnCredential[]> = await response.json();
      
      if (data.success && data.data) {
        return data.data;
      } else {
        throw new Error(data.error || 'Failed to fetch credentials');
      }
    } catch (error) {
      console.error('Error getting user credentials:', error);
      throw error;
    }
  }

  async hasCredentials(userId: string): Promise<boolean> {
    try {
      const response = await authService.apiRequest(`${this.baseUrl}/credentials/${userId}/exists`);
      const data: ApiResponse<boolean> = await response.json();
      
      if (data.success && data.data !== undefined) {
        return data.data;
      } else {
        throw new Error(data.error || 'Failed to check credentials');
      }
    } catch (error) {
      console.error('Error checking credentials:', error);
      throw error;
    }
  }

  async deleteCredential(userId: string, credentialId: string): Promise<{ message: string }> {
    try {
      const response = await authService.apiRequest(`${this.baseUrl}/credentials/${userId}/${credentialId}`, {
        method: 'DELETE',
      });
      const data: ApiResponse<string> = await response.json();
      
      if (data.success) {
        return { message: data.message };
      } else {
        throw new Error(data.error || 'Failed to delete credential');
      }
    } catch (error) {
      console.error('Error deleting credential:', error);
      throw error;
    }
  }
}

const webAuthnService = new WebAuthnService();
export default webAuthnService;

File: services/webauthn-service/src/main/java/com/stockfellow/webauthn/service/WebAuthnService.java
package com.stockfellow.webauthn.service;

import com.stockfellow.webauthn.dto.*;
import com.stockfellow.webauthn.entity.WebAuthnChallenge;
import com.stockfellow.webauthn.entity.KeycloakWebAuthnCredential;
import com.stockfellow.webauthn.repository.WebAuthnChallengeRepository;
import com.webauthn4j.WebAuthnManager;
import com.webauthn4j.data.*;
import com.webauthn4j.data.client.Origin;
import com.webauthn4j.data.client.challenge.DefaultChallenge;
import com.webauthn4j.server.ServerProperty;
import com.webauthn4j.util.Base64UrlUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.time.LocalDateTime;
import java.util.Base64;
import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class WebAuthnService {

    private static final Logger logger = LoggerFactory.getLogger(WebAuthnService.class);

    private final WebAuthnChallengeRepository challengeRepository;
    private final KeycloakService keycloakService;
    private final JwtService jwtService;
    private final KeycloakWebAuthnService keycloakWebAuthnService;
    private final WebAuthnManager webAuthnManager;
    private final SecureRandom secureRandom = new SecureRandom();

    @Value("${webauthn.rpId}")
    private String rpId;

    @Value("${webauthn.rpName}")
    private String rpName;

    @Value("${webauthn.origin}")
    private String origin;

    @Value("${webauthn.timeout:60000}")
    private long timeout;

    public WebAuthnService(WebAuthnChallengeRepository challengeRepository,
            KeycloakService keycloakService,
            JwtService jwtService,
            KeycloakWebAuthnService keycloakWebAuthnService) {
        this.challengeRepository = challengeRepository;
        this.keycloakService = keycloakService;
        this.jwtService = jwtService;
        this.keycloakWebAuthnService = keycloakWebAuthnService;
        this.webAuthnManager = WebAuthnManager.createNonStrictWebAuthnManager();
    }

    public RegistrationStartResponse startRegistration(RegistrationStartRequest request) {
        logger.info("Starting WebAuthn registration for user: {} ({})", request.getUsername(), request.getUserId());

        try {

            if (!keycloakService.userExists(request.getUserId())) {
                throw new RuntimeException("User not found in Keycloak: " + request.getUserId());
            }

            // Generate a cryptographically secure challenge
            byte[] challengeBytes = new byte[32];
            secureRandom.nextBytes(challengeBytes);
            String challenge = Base64UrlUtil.encodeToString(challengeBytes);

            // Store the challenge in database -> expires in 5 min
            WebAuthnChallenge challengeEntity = new WebAuthnChallenge(
                    challenge,
                    request.getUserId(),
                    request.getUsername(),
                    WebAuthnChallenge.ChallengeType.REGISTRATION,
                    LocalDateTime.now().plusMinutes(5));
            challengeRepository.save(challengeEntity);

            // Create the registration response for frontend
            RegistrationStartResponse response = new RegistrationStartResponse();
            response.setChallenge(challenge);
            response.setRpId(rpId);
            response.setRpName(rpName);
            response.setTimeout(timeout);
            response.setAttestation("none");

            response.setUser(new RegistrationStartResponse.UserInfo(
                    Base64UrlUtil.encodeToString(request.getUserId().getBytes()),
                    request.getUsername(),
                    request.getUsername()));

            response.setPubKeyCredParams(new RegistrationStartResponse.PubKeyCredParams[] {
                    new RegistrationStartResponse.PubKeyCredParams("public-key", -7),
                    new RegistrationStartResponse.PubKeyCredParams("public-key", -257)
            });

            response.setAuthenticatorSelection(new RegistrationStartResponse.AuthenticatorSelection(
                    "platform",
                    "required",
                    false));

            logger.info("Successfully generated registration challenge for user: {}", request.getUsername());
            return response;

        } catch (Exception e) {
            logger.error("Failed to start registration for user: {}", request.getUsername(), e);
            throw new RuntimeException("Failed to start registration: " + e.getMessage());
        }
    }

    public ApiResponse<String> completeRegistration(String challenge, RegistrationCompleteRequest request) {
        logger.info("Completing WebAuthn registration for credential: {}", request.getCredentialId());

        try {

            Optional<WebAuthnChallenge> challengeOpt = challengeRepository
                    .findByChallengeAndIsUsedFalseAndExpiresAtAfter(challenge, LocalDateTime.now());

            if (!challengeOpt.isPresent()) {
                return ApiResponse.error("Invalid or expired challenge");
            }

            WebAuthnChallenge challengeEntity = challengeOpt.get();

            challengeEntity.setIsUsed(true);
            challengeRepository.save(challengeEntity);

            // Parse the registration response using WebAuthn4J library
            // This verifies the cryptographic signature and extracts the public key
            byte[] clientDataJSON = Base64UrlUtil.decode(request.getClientDataJSON());
            byte[] attestationObject = Base64UrlUtil.decode(request.getAttestationObject());

            RegistrationRequest registrationRequest = new RegistrationRequest(
                    attestationObject,
                    clientDataJSON);

            RegistrationParameters registrationParameters = new RegistrationParameters(
                    new ServerProperty(new Origin(origin), rpId, new DefaultChallenge(challenge), null),
                    null,
                    false,
                    true);

            RegistrationData registrationData = webAuthnManager.parse(registrationRequest);
            webAuthnManager.validate(registrationData, registrationParameters);

            // Extract and store the credential in Keycloak database

            byte[] credentialId = registrationData.getAttestationObject().getAuthenticatorData()
                    .getAttestedCredentialData().getCredentialId();
            String credentialIdBase64 = Base64.getEncoder().encodeToString(credentialId);

            String publicKeyBase64 = credentialIdBase64;

            keycloakWebAuthnService.storeWebAuthnCredential(
                    challengeEntity.getUserId(),
                    request.getCredentialId(),
                    publicKeyBase64,
                    request.getAuthenticatorName(),
                    0L);

            logger.info("Successfully registered WebAuthn credential for user: {}", challengeEntity.getUsername());
            return ApiResponse.success("Biometric authentication registered successfully");

        } catch (Exception e) {
            logger.error("Failed to complete registration for credential: {}", request.getCredentialId(), e);
            return ApiResponse.error("Registration failed: " + e.getMessage());
        }
    }

    public AuthenticationStartResponse startAuthentication(AuthenticationStartRequest request) {
        logger.info("Starting WebAuthn authentication for user: {}", request.getUsername());

        try {

            String userId = keycloakService.getUserIdByUsername(request.getUsername());
            if (userId == null) {
                throw new RuntimeException("User not found: " + request.getUsername());
            }

            if (!keycloakWebAuthnService.hasWebAuthnCredentials(userId)) {
                throw new RuntimeException("No biometric credentials found for user: " + request.getUsername());
            }

            // Generate challenge
            byte[] challengeBytes = new byte[32];
            secureRandom.nextBytes(challengeBytes);
            String challenge = Base64UrlUtil.encodeToString(challengeBytes);

            // Store challenge
            WebAuthnChallenge challengeEntity = new WebAuthnChallenge(
                    challenge,
                    userId,
                    request.getUsername(),
                    WebAuthnChallenge.ChallengeType.AUTHENTICATION,
                    LocalDateTime.now().plusMinutes(5));
            challengeRepository.save(challengeEntity);

            // Create authentication response
            AuthenticationStartResponse response = new AuthenticationStartResponse();
            response.setChallenge(challenge);
            response.setRpId(rpId);
            response.setTimeout(timeout);
            response.setUserVerification("required"); // Require biometric verification

            List<String> credentialIds = keycloakWebAuthnService.getCredentialIds(userId);
            AuthenticationStartResponse.AllowCredentials[] allowCredentials = credentialIds.stream()
                    .map(credId -> new AuthenticationStartResponse.AllowCredentials(
                            "public-key",
                            credId,
                            new String[] { "internal" } // Platform authenticator
                    ))
                    .toArray(AuthenticationStartResponse.AllowCredentials[]::new);

            response.setAllowCredentials(allowCredentials);

            logger.info("Successfully generated authentication challenge for user: {}", request.getUsername());
            return response;

        } catch (Exception e) {
            logger.error("Failed to start authentication for user: {}", request.getUsername(), e);
            throw new RuntimeException("Failed to start authentication: " + e.getMessage());
        }
    }

    public ApiResponse<AuthenticationResponse> completeAuthentication(String challenge,
            AuthenticationCompleteRequest request) {
        logger.info("Completing WebAuthn authentication for credential: {}", request.getCredentialId());

        try {

            Optional<WebAuthnChallenge> challengeOpt = challengeRepository
                    .findByChallengeAndIsUsedFalseAndExpiresAtAfter(challenge, LocalDateTime.now());

            if (!challengeOpt.isPresent()) {
                return ApiResponse.error("Invalid or expired challenge");
            }

            WebAuthnChallenge challengeEntity = challengeOpt.get();

            String publicKey = keycloakWebAuthnService.getPublicKeyByCredentialId(challengeEntity.getUserId(),
                    request.getCredentialId());
            if (publicKey == null) {
                return ApiResponse.error("Credential not found");
            }

            challengeEntity.setIsUsed(true);
            challengeRepository.save(challengeEntity);

            byte[] credentialId = Base64UrlUtil.decode(request.getCredentialId());
            byte[] clientDataJSON = Base64UrlUtil.decode(request.getClientDataJSON());
            byte[] authenticatorData = Base64UrlUtil.decode(request.getAuthenticatorData());
            byte[] signature = Base64UrlUtil.decode(request.getSignature());

            AuthenticationRequest authRequest = new AuthenticationRequest(
                    credentialId,
                    authenticatorData,
                    clientDataJSON,
                    signature);

            AuthenticationParameters authParameters = new AuthenticationParameters(
                    new ServerProperty(new Origin(origin), rpId, new DefaultChallenge(challenge), null),
                    null,
                    null,
                    false,
                    true);

            // WebAuthn verifies the cryptographic
            // signature
            AuthenticationData authData = webAuthnManager.parse(authRequest);
            webAuthnManager.validate(authData, authParameters);

            KeycloakWebAuthnCredential credential = keycloakWebAuthnService.getWebAuthnCredential(
                    challengeEntity.getUserId(), request.getCredentialId());

            if (credential == null) {
                throw new RuntimeException("Credential not found for user");
            }

            keycloakWebAuthnService.updateSignatureCount(challengeEntity.getUserId(), request.getCredentialId(),
                    authData.getAuthenticatorData().getSignCount());

            String username = keycloakService.getUsernameById(challengeEntity.getUserId());

            String token = jwtService.generateToken(challengeEntity.getUserId(), username);

            AuthenticationResponse authResponse = new AuthenticationResponse(
                    token,
                    challengeEntity.getUserId(),
                    username,
                    jwtService.getExpirationTime());

            logger.info("Successfully authenticated user: {} using WebAuthn", username);
            return ApiResponse.success("Authentication successful", authResponse);

        } catch (Exception e) {
            logger.error("Failed to complete authentication for credential: {}", request.getCredentialId(), e);
            return ApiResponse.error("Authentication failed: " + e.getMessage());
        }
    }

    public List<KeycloakWebAuthnCredential> getUserCredentials(String userId) {
        return keycloakWebAuthnService.getUserCredentials(userId);
    }

    public boolean hasCredentials(String userId) {
        return keycloakWebAuthnService.hasWebAuthnCredentials(userId);
    }

    public boolean deleteCredential(String userId, String credentialId) {
        return keycloakWebAuthnService.deleteCredential(userId, credentialId);
    }

    @Transactional
    public void cleanupExpiredChallenges() {
        challengeRepository.deleteByExpiresAtBefore(LocalDateTime.now());
        challengeRepository.deleteByIsUsedTrue();
    }
}
